### PR TITLE
Handle port-waiting timeout; Catch Python debugging Errors

### PR DESCRIPTION
**Pull request by awschristou**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

While working on some recent changes I found some less-than-ideal debugging handling cases. I put those in a separate PR (here):
* If waiting for the debug port times out, an error is thrown with the string 'timeout' which surfaces in a meaningless showInformation dialog. This has been changed to contain the Error, then message to the output channel as part of the debug experience, that their debug may fail.
* Python local run/debugging has no overall error catch, meaning users have no feedback that things have stopped. A catch was added, and the debug wrapper is now more appropriately removed.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes

- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/codelens/localLambdaRunner.ts [hunk 1]
```diff
@@ -247,12 +247,20 @@ export class LocalLambdaRunner {
         await command.execute()
 
         if (this.localInvokeParams.isDebug) {
-            await waitForDebugPort({
+            const isPortOpen: boolean = await waitForDebugPort({
```
Comment (bryceitoc9) [inline src/shared/codelens/localLambdaRunner.ts]: What causes `waitForDebugPort` to fail? Should that failing just straight-up cancel the debug session?

Comment (awschristou) [reply src/shared/codelens/localLambdaRunner.ts]: Waiting for port has a configurable timeout limit. It fails if the port is not available within that time.

File: src/shared/codelens/localLambdaRunner.ts [hunk 2]
```diff
@@ -653,11 +668,19 @@ async function waitForDebugPort({
...
+    try {
+        await tcpPortUsed.waitUntilUsed(
+            debugPort,
+            SAM_LOCAL_PORT_CHECK_RETRY_INTERVAL_MILLIS,
+            timeoutMillis
+        )
+
+        return true
+    } catch (err) {
+        channelLogger.logger.verbose(`Failed to wait for port ${debugPort} to open: ${err}`)
```
Comment (mpiroc) [inline src/shared/codelens/localLambdaRunner.ts]: What do you think of something like `timed out after ${timeoutMillis}ms waiting for port ${debugPort} to open: ${err}`

File: src/shared/codelens/localLambdaRunner.ts (hunk 1 above)
Comment (awschristou) [reply src/shared/codelens/localLambdaRunner.ts]: Currently the system throws an Error, which we don't suitably handle, so this is a better wrapper around that.

File: src/shared/codelens/pythonCodeLensProvider.ts [hunk 3]
```diff
@@ -209,70 +209,81 @@ export async function initialize({
...
+        } catch (err) {
+            const error = err as Error
+            channelLogger.error(
+                'AWS.error.during.sam.local',
+                'An error occurred trying to run SAM Application locally: {0}',
+                error
+            )
+        } finally {
+            if (lambdaDebugFilePath) {
+                await unlink(lambdaDebugFilePath)
```
Comment (mpiroc) [inline src/shared/codelens/pythonCodeLensProvider.ts]: I think this throws if `lambdaDebugFilePath` does not exist. Should we wrap this line in a try/catch and swallow errors?

Comment (awschristou) [reply src/shared/codelens/pythonCodeLensProvider.ts]: I'll move the unlink call into a separate function and swallow errors to keep this section leaner

File: src/shared/codelens/localLambdaRunner.ts [hunk 4]
```diff
@@ -247,12 +247,20 @@ export class LocalLambdaRunner {
...
                 debugPort: this.debugPort,
                 configuration: this.configuration,
                 channelLogger: this.channelLogger
             })
 
+            if (!isPortOpen) {
+                this.channelLogger.warn(
+                    'AWS.samcli.local.invoke.port.not.open',
+                    // tslint:disable-next-line:max-line-length
+                    'Unable to detect the debug port. The debugger might not succeed in attaching to your SAM Application.'
```
Comment (mpiroc) [inline src/shared/codelens/localLambdaRunner.ts]: Nit: To me, 'detect' reads as "We tried to determine the right port, but couldn't", not "We know the right port, it just isn't open".

**Pull request merged**